funsor/terms.py

```diff
@@ -328,6 +313,14 @@ def pretty(self, maxlen=40):
     def __contains__(self, item):
         raise TypeError
 
+    def _alpha_convert(self, alpha_subs):
+        """
+        Rename bound variables while preserving all free variables.
+        """
+        # Substitute all funsor values.
+        # Subclasses must handle string conversion.
+        return tuple(substitute(v, alpha_subs) for v in self._ast_values)
```

Comment:
Now that the logic has been split up across so many terms, can you add an assertion here checking that `alpha_subs` is only applied to bound variables?  Something like
```py
assert not frozenset(alpha_subs) - self.bound
```

Reply:
Done.